// @base-org/account/dist/core/telemetry/events/provider.js
import { ActionType, AnalyticsEventImportance, ComponentType, logEvent } from '../logEvent.js';
export const logRequestStarted = ({ method, correlationId, isEphemeral = false, }) => {
    logEvent('provider.request.started', {
        action: ActionType.unknown,
        componentType: ComponentType.unknown,
        method,
        signerType: 'base-account',
        correlationId,
        isEphemeral,
    }, AnalyticsEventImportance.high);
};
export const logRequestError = ({ method, correlationId, errorMessage, isEphemeral = false, }) => {
    logEvent('provider.request.error', {
        action: ActionType.error,
        componentType: ComponentType.unknown,
        method,
        signerType: 'base-account',
        correlationId,
        errorMessage,
        isEphemeral,
    }, AnalyticsEventImportance.high);
};
export const logRequestResponded = ({ method, correlationId, isEphemeral = false, }) => {
    logEvent('provider.request.responded', {
        action: ActionType.unknown,
        componentType: ComponentType.unknown,
        method,
        signerType: 'base-account',
        correlationId,
        isEphemeral,
    }, AnalyticsEventImportance.high);
};
export const logGetInjectedProviderError = ({ errorMessage, }) => {
    logEvent('provider.getInjectedProvider.error', {
        action: ActionType.error,
        componentType: ComponentType.unknown,
        method: 'getInjectedProvider',
        signerType: 'base-account',
        errorMessage,
    }, AnalyticsEventImportance.high);
};
