// @base-org/account/dist/store/store.js
import { PACKAGE_VERSION } from '../core/constants.js';
import { createJSONStorage, persist } from 'zustand/middleware';
import { createStore } from 'zustand/vanilla';
const createChainSlice = () => {
    return {
        chains: [],
    };
};
const createKeysSlice = () => {
    return {
        keys: {},
    };
};
const createAccountSlice = () => {
    return {
        account: {},
    };
};
const createSubAccountSlice = () => {
    return {
        subAccount: undefined,
    };
};
const createSubAccountConfigSlice = () => {
    return {
        subAccountConfig: {},
    };
};
const createSpendPermissionsSlice = () => {
    return {
        spendPermissions: [],
    };
};
const createConfigSlice = () => {
    return {
        config: {
            version: PACKAGE_VERSION,
        },
    };
};
/**
 * Factory function to create a store instance.
 * Allows creating either persistent (for regular SDK) or ephemeral (for payment flows) stores.
 */
export function createStoreInstance(options) {
    const { persist: shouldPersist = true, storageName = 'base-acc-sdk.store' } = options ?? {};
    const storeCreator = (...args) => ({
        ...createChainSlice(...args),
        ...createKeysSlice(...args),
        ...createAccountSlice(...args),
        ...createSubAccountSlice(...args),
        ...createSpendPermissionsSlice(...args),
        ...createConfigSlice(...args),
        ...createSubAccountConfigSlice(...args),
    });
    if (shouldPersist) {
        return createStore(persist(storeCreator, {
            name: storageName,
            storage: createJSONStorage(() => localStorage),
            partialize: (state) => {
                // Explicitly select only the data properties we want to persist
                // (not the methods)
                return {
                    chains: state.chains,
                    keys: state.keys,
                    account: state.account,
                    subAccount: state.subAccount,
                    spendPermissions: state.spendPermissions,
                    config: state.config,
                };
            },
        }));
    }
    // Create ephemeral store without persistence
    return createStore(storeCreator);
}
// Global singleton store for backwards compatibility and persistent SDK instances
export const sdkstore = createStoreInstance({ persist: true });
/**
 * Creates store accessor helpers for a given store instance.
 * This allows both the global store and ephemeral stores to use the same API.
 */
export function createStoreHelpers(storeInstance) {
    return {
        subAccountsConfig: {
            get: () => storeInstance.getState().subAccountConfig,
            set: (subAccountConfig) => {
                storeInstance.setState((state) => ({
                    subAccountConfig: { ...state.subAccountConfig, ...subAccountConfig },
                }));
            },
            clear: () => {
                storeInstance.setState({
                    subAccountConfig: {},
                });
            },
        },
        subAccounts: {
            get: () => storeInstance.getState().subAccount,
            set: (subAccount) => {
                storeInstance.setState((state) => ({
                    subAccount: state.subAccount
                        ? { ...state.subAccount, ...subAccount }
                        : { address: subAccount.address, ...subAccount },
                }));
            },
            clear: () => {
                storeInstance.setState({
                    subAccount: undefined,
                });
            },
        },
        spendPermissions: {
            get: () => storeInstance.getState().spendPermissions,
            set: (spendPermissions) => {
                storeInstance.setState({ spendPermissions });
            },
            clear: () => {
                storeInstance.setState({
                    spendPermissions: [],
                });
            },
        },
        account: {
            get: () => storeInstance.getState().account,
            set: (account) => {
                storeInstance.setState((state) => ({
                    account: { ...state.account, ...account },
                }));
            },
            clear: () => {
                storeInstance.setState({
                    account: {},
                });
            },
        },
        chains: {
            get: () => storeInstance.getState().chains,
            set: (chains) => {
                storeInstance.setState({ chains });
            },
            clear: () => {
                storeInstance.setState({
                    chains: [],
                });
            },
        },
        keys: {
            get: (key) => storeInstance.getState().keys[key],
            set: (key, value) => {
                storeInstance.setState((state) => ({ keys: { ...state.keys, [key]: value } }));
            },
            clear: () => {
                storeInstance.setState({
                    keys: {},
                });
            },
        },
        config: {
            get: () => storeInstance.getState().config,
            set: (config) => {
                storeInstance.setState((state) => ({ config: { ...state.config, ...config } }));
            },
        },
    };
}
// Global store with helpers for backwards compatibility
const globalStoreHelpers = createStoreHelpers(sdkstore);
// Re-export global helpers for backwards compatibility
export const subAccountsConfig = globalStoreHelpers.subAccountsConfig;
export const subAccounts = globalStoreHelpers.subAccounts;
export const spendPermissions = globalStoreHelpers.spendPermissions;
export const account = globalStoreHelpers.account;
export const chains = globalStoreHelpers.chains;
export const keys = globalStoreHelpers.keys;
export const config = globalStoreHelpers.config;
export const store = {
    ...sdkstore,
    ...globalStoreHelpers,
    persist: sdkstore.persist,
};
