// @base-org/account/dist/interface/builder/core/EphemeralBaseAccountProvider.js
import { Communicator } from '../../../core/communicator/Communicator.js';
import { CB_WALLET_RPC_URL } from '../../../core/constants.js';
import { standardErrorCodes } from '../../../core/error/constants.js';
import { standardErrors } from '../../../core/error/errors.js';
import { serializeError } from '../../../core/error/serialize.js';
import { ProviderEventEmitter, } from '../../../core/provider/interface.js';
import { hexStringFromNumber } from '../../../core/type/util.js';
import { EphemeralSigner } from '../../../sign/base-account/EphemeralSigner.js';
import { createStoreInstance } from '../../../store/store.js';
import { fetchRPCRequest } from '../../../util/provider.js';
import { withMeasurement } from './withMeasurement.js';
/**
 * EphemeralBaseAccountProvider is a provider designed for single-use payment flows.
 *
 * Key differences from BaseAccountProvider:
 * 1. Creates its own isolated store instance (no persistence, no global state pollution)
 * 2. Uses EphemeralSigner with the isolated store to prevent concurrent operation interference
 * 3. Cleanup clears the entire ephemeral store instance
 * 4. Optimized for one-shot operations like pay() and subscribe()
 *
 * This prevents:
 * - Race conditions when multiple ephemeral payment flows run concurrently
 * - KeyManager interference (each instance has its own isolated keys)
 * - Memory leaks (store instance is garbage collected after cleanup)
 */
export class EphemeralBaseAccountProvider extends ProviderEventEmitter {
    communicator;
    signer;
    ephemeralStore;
    constructor({ metadata, preference: { walletUrl, ...preference }, }) {
        super();
        this.communicator = new Communicator({
            url: walletUrl,
            metadata,
            preference,
        });
        // Create an isolated ephemeral store for this provider instance
        // persist: false means no localStorage persistence
        this.ephemeralStore = createStoreInstance({ persist: false });
        this.signer = new EphemeralSigner({
            metadata,
            communicator: this.communicator,
            callback: this.emit.bind(this),
            storeInstance: this.ephemeralStore,
        });
    }
    request = withMeasurement({ isEphemeral: true }, async (args) => {
        try {
            // For ephemeral providers, we only support a subset of methods
            // that are needed for payment flows
            switch (args.method) {
                case 'wallet_sendCalls':
                case 'wallet_sign': {
                    try {
                        await this.signer.handshake({ method: 'handshake' }); // exchange session keys
                        const result = await this.signer.request(args); // send diffie-hellman encrypted request
                        return result;
                    }
                    finally {
                        await this.signer.cleanup(); // clean up (rotate) the ephemeral session keys
                    }
                }
                case 'wallet_getCallsStatus': {
                    const result = await fetchRPCRequest(args, CB_WALLET_RPC_URL);
                    return result;
                }
                case 'eth_accounts': {
                    return [];
                }
                case 'net_version': {
                    const result = 1; // default value
                    return result;
                }
                case 'eth_chainId': {
                    const result = hexStringFromNumber(1); // default value
                    return result;
                }
                default: {
                    throw standardErrors.provider.unauthorized(`Method '${args.method}' is not supported by ephemeral provider. Ephemeral providers only support: wallet_sendCalls, wallet_sign, wallet_getCallsStatus`);
                }
            }
        }
        catch (error) {
            const { code } = error;
            if (code === standardErrorCodes.provider.unauthorized) {
                await this.disconnect();
            }
            return Promise.reject(serializeError(error));
        }
    });
    async disconnect() {
        // Cleanup ephemeral signer state and its isolated store
        await this.signer.cleanup();
        // Note: The ephemeral store instance will be garbage collected
        // when this provider instance is no longer referenced
        this.emit('disconnect', standardErrors.provider.disconnected('User initiated disconnection'));
    }
    isBaseAccount = true;
}
